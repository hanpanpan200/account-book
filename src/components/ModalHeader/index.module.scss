.container {
  display: flex;
  align-items: center;
  width: 100vw;
  height: 50px;
  justify-content: center;

  span {
    font-size: 16px;
    font-weight: 400;
  }
  img {
    width: 28px;
    height: 28px;
  }
  .left {
    position: absolute;
    left: 10px;
  }
  .right {
    position: absolute;
    right: 10px;
  }
}