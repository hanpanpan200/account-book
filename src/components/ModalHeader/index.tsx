import React from 'react';
import classNames from 'classnames';

import styles from './index.module.scss';

const cx = classNames.bind(styles);

export enum TitlePosition {
  Left,
  Center,
}

interface Props {
  title: string
  leftText?: string
  rightText?: string
  leftIconSource?: string
  rightIconSource?: string
  onLeftClick?: () => void
  onRightClick?: () => void
  titlePosition?: TitlePosition
}

const ModalHeader: React.FC<Props> = ({ title, titlePosition, leftText, rightText, leftIconSource, rightIconSource, onLeftClick, onRightClick }) => {
  const titleStyle = cx({
    [styles.leftTitle]: titlePosition === TitlePosition.Left,
    [styles.centralTitle]: !titlePosition || titlePosition === TitlePosition.Center,
  })
  return (
    <div className={styles.container}>
      {leftText && <span className={styles.left} onClick={onLeftClick}>{ leftText }</span> }
      {leftIconSource && <img className={styles.left} src={leftIconSource} onClick={onLeftClick} alt="modalHeaderLeftIcon"/>}
      <span className={titleStyle}>{title}</span>
      {rightText && <span className={styles.right} onClick={onRightClick}>{ rightText }</span> }
      {rightIconSource && <img className={styles.right} src={rightIconSource} onClick={onRightClick} alt="modalHeaderRightIcon"/>}
    </div>
  )
}

export default ModalHeader;