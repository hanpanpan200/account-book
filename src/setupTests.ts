// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import { fireEvent } from '@testing-library/react';

export const fireClickEvent = (element: Document | Element | Window | Node) => {
  fireEvent(element, new MouseEvent('click', {
    bubbles: true,
    cancelable: true
  }));
}