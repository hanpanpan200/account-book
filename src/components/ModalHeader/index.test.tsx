import React from 'react';
import { render } from '@testing-library/react';
import ModalHeader, { TitlePosition } from './index';
import { fireClickEvent } from '../../setupTests';

describe('ModalHeader', () => {
  const mockedOnLeftClicked = jest.fn();
  const mockedOnRightClicked = jest.fn();

  it('should render correctly when left and right button are provided as string and titlePosition is Left', async () => {
    const { container, findByText } = render(
      <ModalHeader
        title='Modal Header Title'
        titlePosition={TitlePosition.Left}
        leftText='Left Button'
        rightText='Right Button'
        onLeftClick={mockedOnLeftClicked}
        onRightClick={mockedOnRightClicked}
      />);
    expect(container).toMatchSnapshot();
    const leftText = await findByText('Left Button');
    fireClickEvent(leftText);
    expect(mockedOnLeftClicked).toBeCalled();
    const rightText = await findByText('Right Button');
    fireClickEvent(rightText);
    expect(mockedOnRightClicked).toBeCalled();
  });
  it('should render correctly when left and right button are provided as icon and titlePosition is Center', async () => {
    const { container, findByAltText } = render(
      <ModalHeader
        title='Modal Header Title'
        titlePosition={TitlePosition.Center}
        leftIconSource='leftIcon.svg'
        rightIconSource='rightIcon.svg'
        onLeftClick={mockedOnLeftClicked}
        onRightClick={mockedOnRightClicked}
      />);
    expect(container).toMatchSnapshot();
    const leftIcon = await findByAltText('modalHeaderLeftIcon');
    fireClickEvent(leftIcon);
    expect(mockedOnLeftClicked).toBeCalled();
    const rightIcon = await findByAltText('modalHeaderRightIcon');
    fireClickEvent(rightIcon);
    expect(mockedOnRightClicked).toBeCalled();
  });
});
