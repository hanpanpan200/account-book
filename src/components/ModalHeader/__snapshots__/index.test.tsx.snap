// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalHeader should render correctly when left and right button are provided as icon and titlePosition is Center 1`] = `
<div>
  <div
    class="container"
  >
    <img
      alt="modalHeaderLeftIcon"
      class="left"
      src="leftIcon.svg"
    />
    <span
      class="centralTitle"
    >
      Modal Header Title
    </span>
    <img
      alt="modalHeaderRightIcon"
      class="right"
      src="rightIcon.svg"
    />
  </div>
</div>
`;

exports[`ModalHeader should render correctly when left and right button are provided as string and titlePosition is Left 1`] = `
<div>
  <div
    class="container"
  >
    <span
      class="left"
    >
      Left Button
    </span>
    <span
      class="leftTitle centralTitle"
    >
      Modal Header Title
    </span>
    <span
      class="right"
    >
      Right Button
    </span>
  </div>
</div>
`;
